crates/forge/src/compatibility_check.rs

```diff
@@ -0,0 +1,177 @@
+use anyhow::{anyhow, Context, Result};
+use regex::Regex;
+use semver::Version;
+use std::process::Command;
+
+type VersionParser<'a> = dyn Fn(&str) -> Result<Version> + 'a;
+
+pub struct Requirement<'a> {
+    pub name: String,
+    pub command: String,
+    pub version_parser: Box<VersionParser<'a>>,
+    pub helper_text: String,
+    pub minimal_version: Version,
+}
+
+pub struct RequirementsChecker<'a> {
+    output_on_success: bool,
+    requirements: Vec<Requirement<'a>>,
+}
+
+impl<'a> RequirementsChecker<'a> {
+    pub(crate) fn new(output_on_success: bool) -> Self {
+        Self {
+            output_on_success,
+            requirements: Vec::new(),
+        }
+    }
+
+    pub fn add_requirement(&mut self, requirement: Requirement<'a>) {
+        self.requirements.push(requirement);
+    }
+
+    pub fn validate(&self) -> Result<()> {
+        let (validation_output, is_valid) = self.validate_with_output()?;
+
+        if !is_valid || self.output_on_success {
+            println!("{validation_output}");
+            if is_valid {
+                return Ok(());
+            }
+            return Err(anyhow!("Requirements not satisfied"));
+        }
+
+        Ok(())
+    }
+
+    fn validate_with_output(&self) -> Result<(String, bool)> {
+        let mut validation_output = "Validating requirements\n\n".to_string();
+        let mut is_valid = true;
+
+        for requirement in &self.requirements {
+            let raw_version = get_raw_version(&requirement.name, &requirement.command)?;
+            let version = (requirement.version_parser)(&raw_version)?;
+            let valid = version >= requirement.minimal_version;
+            let command_output = if valid {
+                format!("✅ {} {}", requirement.name, version)
+            } else {
+                is_valid = false;
+                format!(
+                    "❌ {} Version {} doesn't satisfy minimum {}\n{}",
+                    requirement.name, version, requirement.minimal_version, requirement.helper_text
+                )
+            };
+
+            validation_output += command_output.as_str();
+            validation_output += "\n";
+        }
+
+        Ok((validation_output, is_valid))
```

Comment:
`is_valid` variable indicates whether all requirements are valid, while `valid` represents the validity of a single requirement. It should be rather `all_valid`/`all_requirements_valid` and `valid`/`is_valid` (smth similar) to make it more clarified.